internal/expr/parser.go

```diff
@@ -532,18 +539,43 @@ func (p *Parser) parseOutputExpression() (*outputPart, bool, error) {
 	return nil, false, nil
 }
 
-// parseInputExpression parses an input expression of the form "$Type.name".
+// parseInputExpression parses an input expression of the form "Type.name".
 func (p *Parser) parseInputExpression() (*inputPart, bool, error) {
 	cp := p.save()
 
+	if fn, ok, err := p.parseSourceType(); ok {
+		return &inputPart{sourceType: fn, raw: p.input[cp.pos:p.pos]}, true, nil
+	} else if err != nil {
+		return nil, false, err
+	}
+
+	cp.restore()
+	return nil, false, nil
+}
+
+func (p *Parser) parseSourceType() (fullName, bool, error) {
 	if p.skipByte('$') {
 		if fn, ok, err := p.parseGoFullName(); ok {
-			if fn.name == "*" {
-				return nil, false, fmt.Errorf(`asterisk not allowed in input expression "$%s"`, fn)
-			}
-			return &inputPart{sourceType: fn, raw: p.input[cp.pos:p.pos]}, true, nil
+			return fn, true, nil
 		} else if err != nil {
+			return fullName{}, false, err
+		}
+	}
+	return fullName{}, false, nil
+}
+
+func (p *Parser) parseInExpression() (*inPart, bool, error) {
+	cp := p.save()
+
+	if p.skipString("IN") {
+		p.skipBlanks()
+		if types, ok, err := p.parseList((*Parser).parseSourceType); err != nil {
 			return nil, false, err
+		} else if ok {
+			return &inPart{
+				types: types,
+				raw:   p.input[cp.pos:p.pos],
+			}, true, nil
```

Comment:
Per previous reviews, let's please not have this pattern as a rule. This line is just:
```
return &inPart{types: types, raw: p.input[cp.pos:p.pos]}, true, nil
```
If that's not readable enough, then the better alternative is:
```
part := &inPart{types: types, raw: p.input[cp.pos:p.pos]}
return part, true, nil

Reply:
Changed.